### PR TITLE
Use store namespacing

### PR DESCRIPTION
**Pull request by JensGM**

### CODE REVIEW

**Review by achaikou — COMMENTED** [review]
I almost can't believe I am saying this, but I like this patch!!! :smile: :smile: :smile:  It's _**beautiful**_
Almost clear new concepts, no :scream_cat: and finally getting rid of FutureValueNodes!

File: xun/functions/function.py [hunk 1]
```diff
@@ -169,25 +198,26 @@ def callable(self, extra_globals=None):
...
         new_globals = {
-            name: value for name, value in fimg.desc.globals.items()
+            name: value for name, value in self.desc.globals.items()
             if not isinstance(value, Function)
         }
         if extra_globals is not None:
             new_globals.update(extra_globals)
 
         f.globals = new_globals
+        f.hash = self.hash
```
[the remark this answers is not in the document]
Comment (JensGM) [reply xun/functions/function.py:220]: Function are not picklable, FunctionImage runnable produced by the Function

File: xun/functions/graph.py [hunk 2]
```diff
@@ -43,6 +43,12 @@ def __init__(self, function_name, *args, **kwargs):
         self.args = args
         self.kwargs = kwargs
 
+    def __copy__(self):
+        raise CopyError('Cannot copy value')
+
+    def __deepcopy__(self, memo=None):
+        raise CopyError('Cannot copy value')
+
```
[the remark this answers is not in the document]
Comment (JensGM) [reply xun/functions/graph.py:56]: Added a comment

File: xun/functions/store/store_accessor.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+from .. import CallNode
+
+
+class StoreAccessor:
```
[the remark this answers is not in the document]
Comment (JensGM) [reply xun/functions/store/store_accessor.py:4]: Added

File: xun/functions/store/store_accessor.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        """
+        Given a call, replace any FutureValueNodes with values from the store.
+
+        Parameters
+        ----------
+        call : CallNode
+
+        Returns
+        CallNode
+            Call with FutureValueNodes replaced by the value they represent
```
[the remark this answers is not in the document]
Comment (JensGM) [reply xun/functions/store/store_accessor.py]: Updated the function

File: xun/functions/transformations.py [hunk 5]
```diff
@@ -497,7 +496,7 @@ def map(self, nodes):
             transformed = (self.visit(copy.deepcopy(node)) for node in nodes)
             return [node for node in transformed if node is not None]
 
-    class Call2Future(NodeMapper):
+    class Call2CallNode(NodeMapper):
```
[the remark this answers is not in the document]
Comment (JensGM) [reply xun/functions/transformations.py:502]: Added some comments on the usage of the classes, the classes them selves are uninteresting and exist only as artifacts of the ast interface

File: xun/functions/graph.py [hunk 6]
```diff
@@ -29,6 +29,11 @@ class CallNode:
     store. When a call is executed, the result is stored in the program store
     using the CallNode as key.
 
+    CallNodes are used as sentinel values during scheduling. To protect against
+    attempted modification during scheduling, copying of CallNodes is
+    disallowed. This is because the value a CallNode represents is not known
+    until execution and can therefore not be used.
```
Comment (achaikou) [inline xun/functions/graph.py:35]: [rant starts]

I will never understand your no-copy explanation :D
You keep explaining it, I keep being confused :D 

Like in this explanation my first thought goes to "what the problem with value not known until execution"? OK, I'll copy the node and then resolve it the same way as original one? I see precisely no problem here.
If we include your "copy is pickling" explanation, then it doesn't work either. CallNodes can be pickled and are pickled, so what?

[rant ends]

Please see if there are some points in this rant and docs can be made even better :)

File: xun/functions/transformations.py [hunk 7]
```diff
@@ -540,7 +545,7 @@ def visit_Call(self, node):
             store_accessor_load_call = ast.Call(
                 func=store_accessor_load_func,
                 args=[call_node],
-                keywords=[],
+                keywords=[ast.keyword('hash', hash_expr)],
```
Comment (achaikou) [inline xun/functions/transformations.py:548]: [github is stupid, so new comment]
Can you test that correct hash is used or is it too specific case?
If I remove the keyword thing now, nothing fails.

File: xun/tests/test_functions.py [hunk 8]
```diff
@@ -189,6 +191,25 @@ def test_blueprint_is_picklable():
...
+        pass
+
+    @xun.function()
+    def g():
+        with ...:
+            a = f()
+            b = use(a)
+        return b
+
+    with pytest.raises(CopyError):
```
[the remark this answers is not in the document]
Comment (achaikou) [reply xun/tests/test_functions.py:209]: Still. As a user I am quite confused about what I did wrong.
To me it looks like quicksort example? I count value, I use it. Why do I get a *copy* error??? I didn't copy anything..

[I clearly do not understand xun :D Which is probably good. I will be your stupid user :D ]

File: xun/functions/graph.py (hunk 6 above)
Comment (achaikou) [reply xun/functions/graph.py:35]: Of course we know what they are!!!!
They exist and we know that they don't know what they are before execution, but nothing should stop us from copying this "I do no not know what I am" to another place and then just finding out during execution what it is separately for both nodes!

**Review by achaikou — APPROVED** [review]
Fiiiiine. If you wan to ignore poor formatter, merge while I am kind :D 

P.S. It's your and only your fault that I now pronounce "fiiine" with your intonation!!!

**Pull request merged**